### ISSUE TITLE
question about failing of multiple chaincode upgrades in parallel

### ISSUE THREAD

**Issue by LudoNieuw** [APGGroeiFabriek/PIVT#43]
Hi there raft@git and other community members.

As a proud devops engineer that has had the please to work with raft@git on this, I cannot start without noting that I love this project and the ease it provides for running HLF in a manageable way. We have automated the steps in this repo using Azure DevOps, deploying to Azure and it runs like a charm. 

First, I want to share some of our experience that might help the wider community, and secondly, I have a question regarding parallel upgrading of chaincodes.

**1. Because it may be useful to other the following:**
Recently, we have stumbled into issues with Azure Kubernetes Service (AKS) that might have a broader scope than just AKS. The issues concerned really high IOPS values on our nodes that run the hyperledger pods when adding an organisation to a running network. This led in some instances to Nodes crashing and not recovering. As a workaround, we increased OS disk size for the nodes, which gave us a higher IOPS ceiling and has been working for us since. The real challenge is to spread the hyperledger fabric pods evenly across all available nodes. The anti-affinity as implemented in this project does work, but all peer- and orderer pods somehow end up in only two of the 4 available nodes: one copy of the pods in node X and their replica twin in node Y (we have 2 replicas per pods by default). When we have more nodes available, I want to spread the pods across all of them, to further decrease IOPS pressure on the Nodes. Any ideas of how to implement that, are welcome :)

**2. The question I have**
We have 6 chaincode CD pipelines, each of which performs an upgrade chaincode flow for separate chaincode projects. We have some deployments where we set up all of those chaincode flows to run in parallel. However, since recently this causes very long upgrade times, up to which Azure Devops cancels the deployments. While Azure Devops cancels them, with the help of argo cli, I can see the chaincode flows are still running. Install chaincode seems to work work fine, but at instantiate it fails. When I run chaincode upgrades consecutively, they run well. I'm wondering if this may be due to the orderers being too busy, or is there some other aspect that I'm missing here? Any advice or leads would be most welcome.

This is typical error I see in the failing instantiate tries (from `argo logs hlf-chaincodes-l9ptm-1924696291`):
```
Error: endorsement failure during invoke. response: status:500 message:"failed to execute transaction fc1840d9e00c5427253822f7bdb88b84463a5e9c1fa7226e69b7737fa4e04ca4: [channel public] could not launch chaincode info:18999: timeout expired while starting chaincode info:18999 for transaction"
```

**Comment by raftAtGit** [comment]
Thanks for the lovely words Ludo :)

1. Maybe one option is, assigning certain peer/orderer orgs to certain Nodes in the K8S cluster using [`NodeAffinity`](https://kubernetes.io/docs/concepts/configuration/assign-pod-node/#affinity-and-anti-affinity)

For example, one possible configuration will be:
```
OrdererOrgs:
  - Name: Groeifabriek
    Nodes:
      - aks-agentpool-30800063-0
      - aks-agentpool-30800063-1
      - aks-agentpool-30800063-2
   ...
PeerOrgs:
  - Name: Karga
    Nodes:
      - aks-agentpool-30800063-0
      - aks-agentpool-30800063-1
   ...
  - Name: Nevergreen
    Nodes:
      - aks-agentpool-30800063-2
      - aks-agentpool-30800063-3
   ...
```

And the `hlf-kube` chart will use this configuration if defined, and create relevant `NodeAffinity` specs.

Manual configuration can be a pain, so this configuration can be created on the fly with some script using output of `kubectl get node`

PS: 
> we have 2 replicas per pods by default

They are not replicas but separate peer pods created by separate `StatefulSets` ;)

2. This sounds like a Fabric bug similar to this one which I reported long time ago:
https://jira.hyperledger.org/browse/FAB-13471

I guess you should consider opening another Jira ticket.

Meanwhile, as a workaround, you can run the chaincode CD pipelines in a single agent Azure DevOps pool, this will force them run sequential.

**Comment by LudoNieuw** [comment]
Thanks for the support and detailed response! Cool functionality that defining `Node` picks up the NodeAffinity! I hope to find some time to try this out soon. I tried to define the steps of how to automate this. Getting the node names is straightforward, and probably with some yq command, I could script those names to be put at the right place or the tamplate, but to dynamcally choose which node is put at which peer's statefulset definition, requires some deeper thought...

For the chaincode-suggestion, I followed your workaround suggestion and now have one agent pool dedicated for all chaincode CD pipelines with just one agent. Works like a charm and seems to be much faster than when parallel running was actually working successfully. I haven't run many chaincode CDs since, but this workaround looks more promising than getting the parallel install fixed. If indeed running times are much lower, I'll report back on this as others will benefit to know.

Cheers!
